src/main/java/com/es/phoneshop/model/product/DefaultCartService.java

```diff
@@ -0,0 +1,59 @@
+package com.es.phoneshop.model.product;
+
+import com.es.phoneshop.cart.Cart;
+import com.es.phoneshop.cart.CartItem;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class DefaultCartService implements CartService {
+    private static final String CART_SESSION_ATTRIBUTE = DefaultCartService.class.getName() + "cart";
+
+    private Cart cart = new Cart();
+    private ProductDao productDao;
+
+    private DefaultCartService() {
+        this.productDao = ArrayListProductDao.getInstance();
+    }
+
+    private static class SingletonHelper {
+        private static final DefaultCartService INSTANCE = new DefaultCartService();
+    }
+
+    public static DefaultCartService getInstance() {
+        return DefaultCartService.SingletonHelper.INSTANCE;
+    }
+
+    @Override
+    public synchronized Cart getCart(HttpServletRequest request) {
+        Cart cart = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (cart == null) {
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, cart = new Cart());
+        }
+        return cart;
+    }
+
+    @Override
+    public synchronized void add(Cart cart, Long productId, int quantity) throws OutOfStockException {
```

Comment:
Please refactor this method: 1) Change general approach: add setter CartItem::setQuantity and use it instead of removing and adding new objects 2) removeCartItem is bad name (even in current code), for new logic smth like targetCartItem is better 3) Use stream api to filter needed cart item instead of for cycle